### PR TITLE
feat: support skipping legacy code CleanupNetwork in AKS Windows CSE

### PR DESCRIPTION
**Pull request by ShiqianTao**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

- kube-proxy upstream has the code for [waiting azure network](https://github.com/kubernetes/kubernetes/blob/11785bb815d58eb553be3a1fa305464c35d860cc/pkg/proxy/winkernel/proxier.go#L217-L225) which supported [1.22+](https://github.com/kubernetes/kubernetes/commit/d5d9327351de99e6068f62efb388a6ede38d944e)
- kube-proxy upstream has the code for [cleaning HNS policy](https://github.com/kubernetes/kubernetes/blob/fc7325fbec0fe379694e58b82e09f99f166c5e6d/pkg/proxy/winkernel/proxier.go#L685-L686) which supported [1.21+](https://github.com/kubernetes/kubernetes/commit/f3b9e8b10524abfddc52e06eb9a0040d2f1c09df)
- Retain HNS network which confirmed by Windows team and Upstream team
- Reuse toggle `NotRebootWindowsNode` to control it by PR https://github.com/Azure/AgentBaker/pull/4103

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

**Comment by marosset** [comment]
LGTM

File: staging/cse/windows/provisioningscripts/cleanupnetwork.ps1 [hunk 1]
```diff
@@ -1,37 +1,42 @@
...
-    Write-Host "Cleaning up old HNS network found"
-    Remove-HnsNetwork $hnsNetwork
-    Start-Sleep 10
+        Write-Host "Cleaning up old HNS network found"
+        Remove-HnsNetwork $hnsNetwork
+        Start-Sleep 10
+    }
+} else {
+    Write-Host "Skipping network legacy code: Remove-HnsNetwork"
 }
```
Comment (AbelHu) [inline staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:41]: Could we still remove those json files of azurecni if we do not remove hnsnetwork?

File: staging/cse/windows/kubernetesfunc.ps1 [hunk 2]
```diff
@@ -105,6 +105,7 @@ function Write-KubeClusterConfig {
         HNSRemediator       = @{
             IntervalInMinutes = $Global:HNSRemediatorIntervalInMinutes;
         };
+        CleanupHnsNetwork = $global:CleanupHnsNetwork;
```
Comment (AbelHu) [inline staging/cse/windows/kubernetesfunc.ps1:108]: Suggest using `IsCleanupHnsNetwork`

File: staging/cse/windows/provisioningscripts/cleanupnetwork.ps1 [hunk 3]
```diff
@@ -1,37 +1,42 @@
...
 $global:NetworkPlugin = $Global:ClusterConfiguration.Cni.Name
 $global:HNSModule = "c:\k\hns.v2.psm1"
 
 ipmo $global:HNSModule
 
-$networkname = $global:NetworkMode.ToLower()
-if ($global:NetworkPlugin -eq "azure") {
-    $networkname = "azure"
-}
+if ($CleanupHnsNetwork) { # CleanupHnsNetwork is legacy code
```
Comment (AbelHu) [inline staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:10]: Can we test this change in below scenarios?
1. kill hns service
2. kill containerd
3. restart kubelet
4. restart the node
5. ...

File: staging/cse/windows/provisioningscripts/kubeletstart.ps1 [hunk 4]
```diff
@@ -73,14 +74,19 @@ Stop-Service kubeproxy
...
-    # Find if network created by CNI exists, if yes, remove it
-    # This is required to keep the network non-persistent behavior
-    # Going forward, this would be done by HNS automatically during restart of the node
-    & "c:\k\cleanupnetwork.ps1"
-
-    # Restart Kubeproxy, which would wait, until the network is created
-    # This was fixed in 1.15, workaround still needed for 1.14 https://github.com/kubernetes/kubernetes/pull/78612
-    Restart-Service Kubeproxy -Force
+    if ($IsCleanupHnsNetwork) { # CleanupHnsNetwork is legacy code
+    {
```
Comment (junjiezhang1997) [inline staging/cse/windows/provisioningscripts/kubeletstart.ps1:78]: It seems that this left parenthesis is redundant?

**Review by daschott — CHANGES_REQUESTED** [review]
Issue is often observed on initial provisioning of Node by customer. If we wish to preserve the cleanupnetwork.ps1 with HNS removal for initial provisioning, let's update the cleanupnetwork.ps1 as follows:
```
if ($hnsNetwork) {
.... # Remove containers, etc.
Write-Host "Cleaning up old HNS network found"
Get-HnsEndpoint | Remove-HnsEndpoint
Remove-HnsNetwork $hnsNetwork
# Count actively reserved port pools
$countPortPools = 100
while ($countPortPools -ne 0) {
  $countPortPools = (Invoke-HnsRequest -Method "GET" -Type "portpools" | Select-Object -ExpandProperty PortPoolAllocations | Where-Object -FilterScript {$_.RequiresHostPortReservation -eq "True" -and $_.PortRanges} | Select-Object -ExpandProperty PortRanges).Count
  Write-Host "Waiting for HNS to release held port pools..."
  Start-Sleep 10
}
Write-Host "Cleaning up old HNS network completed"
 ```
 NOTE: You will need updated [hns.v2.psm1](https://github.com/microsoft/SDN/blob/master/Kubernetes/windows/hns.v2.psm1#L2559) for `Invoke-HnsRequest -Method "GET" -Type "portpools"`
 
 Alternatively, we can change `Start-Sleep 10` to `Start-Sleep 240`. That should give enough time. But 10s may not guarantee sufficient time as we have seen customers environment ports may still not be released from TCP/IP for large number of hostPorts.

**Comment by daschott** [comment]
cc @sbangari

File: staging/cse/windows/provisioningscripts/cleanupnetwork.ps1 [hunk 5]
```diff
@@ -30,8 +32,20 @@ if ($hnsNetwork) {
...
-    Start-Sleep 10
+    # Count actively reserved port pools
+    $countPortPools = 1
+    $maxIteration=6 # $maxIteration=23
+    For ($i=1; ($i -le $maxIteration -and $countPortPools -ne 0); $i++) {
+      $countPortPools = (Invoke-HnsRequest -Method "GET" -Type "portpools" | Select-Object -ExpandProperty PortPoolAllocations | Where-Object -FilterScript {$_.RequiresHostPortReservation -eq "True" -and $_.PortRanges} | Select-Object -ExpandProperty PortRanges).Count
+      Write-Host "Waiting for HNS to release $countPortPools held port pools ($i/$maxIteration)..."
+      Start-Sleep 10
+    }
+    # Guarantee (10s <= wait time <= 70s) between HNS removal vs. creation for safety.
```
Comment (daschott) [inline staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:47]: I believe the current code is max 60s (not 70s?) but either way lgtm

File: staging/cse/windows/provisioningscripts/kubeproxystart.ps1 [hunk 6]
```diff
@@ -50,11 +51,15 @@ if ($Global:ClusterConfiguration.Kubernetes.Kubeproxy.ConfigArgs) {
...
-$hnsNetwork = Get-HnsNetwork | ? Name -EQ $KubeNetwork
-while (!$hnsNetwork) {
-    Write-Host "$(Get-Date -Format o) Waiting for Network [$KubeNetwork] to be created . . ."
-    Start-Sleep 10
+if ($IsCleanupHnsNetwork) { # CleanupHnsNetwork is legacy code
     $hnsNetwork = Get-HnsNetwork | ? Name -EQ $KubeNetwork
+    while (!$hnsNetwork) {
+        Write-Host "$(Get-Date -Format o) Waiting for Network [$KubeNetwork] to be created . . ."
+        Start-Sleep 10
+        $hnsNetwork = Get-HnsNetwork | ? Name -EQ $KubeNetwork
```
Comment (daschott) [inline staging/cse/windows/provisioningscripts/kubeproxystart.ps1:59]: Assuming IsCleanupHnsNetwork set to false, this code will never run, right? We should not be removing HNS network on kube-proxy restart. Kubeproxy only manages policylists. Kube-proxy will delete policylists itself when it is started, so I don't see any reason to delete any HNS resources. It should never be run imho.

File: staging/cse/windows/provisioningscripts/kubeproxystart.ps1 [hunk 7]
```diff
@@ -81,13 +86,15 @@ if ($featureGateArgs -ne "") {
...
-Import-Module $global:HNSModule
-# Workaround for https://github.com/kubernetes/kubernetes/pull/68923 in < 1.14,
-# and https://github.com/kubernetes/kubernetes/pull/78612 for <= 1.15
-Get-HnsPolicyList | Remove-HnsPolicyList
+if ($IsCleanupHnsNetwork) { # CleanupHnsNetwork is legacy code
+    # cleanup the persisted policy lists
+    Import-Module $global:HNSModule
+    # Workaround for https://github.com/kubernetes/kubernetes/pull/68923 in < 1.14,
+    # and https://github.com/kubernetes/kubernetes/pull/78612 for <= 1.15
+    Get-HnsPolicyList | Remove-HnsPolicyList
```
Comment (daschott) [inline staging/cse/windows/provisioningscripts/kubeproxystart.ps1:94]: Kube-proxy will delete policylists itself when it is started, so I don't see any reason to delete any HNS resources. It should never be run imho so I am glad you disabled this.

File: staging/cse/windows/provisioningscripts/cleanupnetwork.ps1 [hunk 8]
```diff
@@ -30,8 +34,27 @@ if ($hnsNetwork) {
...
 
     Write-Host "Cleaning up old HNS network found"
-    Remove-HnsNetwork $hnsNetwork
-    Start-Sleep 10
+
+    if (-not $IsCleanupHnsNetwork) {
+        # Original code
+        Remove-HnsNetwork $hnsNetwork
+        Start-Sleep 10
+    } else {
```
Comment (tamilmani1989) [inline staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:42]: should always remove hnsendpoint first and then hnsnetwork.. we can keep that part common

File: staging/cse/windows/debug/hns.v2.psm1 [hunk 9]
```diff
@@ -2556,7 +2556,7 @@ function Invoke-HnsRequest
     (
         [ValidateSet('GET', 'POST', 'DELETE')]
         [parameter(Mandatory=$true)] [string] $Method,
-        [ValidateSet('networks', 'endpoints', 'activities', 'policylists', 'endpointstats', 'plugins', 'namespaces', 'globals', 'endpointaddresses')]
+        [ValidateSet('networks', 'endpoints', 'activities', 'policylists', 'endpointstats', 'plugins', 'namespaces', 'globals', 'endpointaddresses', 'portpools')]
```
Comment (AbelHu) [inline staging/cse/windows/debug/hns.v2.psm1:2559]: This file is synced so we should not update it

File: staging/cse/windows/kubernetesfunc.ps1 [hunk 10]
```diff
@@ -105,6 +105,8 @@ function Write-KubeClusterConfig {
         HNSRemediator       = @{
             IntervalInMinutes = $Global:HNSRemediatorIntervalInMinutes;
         };
+        # IsSkipCleanupNetwork = $global:IsSkipCleanupNetwork;
+        IsSkipCleanupNetwork = $true;
```
Comment (AbelHu) [inline staging/cse/windows/kubernetesfunc.ps1:109]: Why changing it to true?

File: staging/cse/windows/provisioningscripts/cleanupnetwork.ps1 [hunk 11]
```diff
@@ -1,5 +1,17 @@
+# The cleanup process for `HNS Network` and `CNI data` should be consistent.
+
 $Global:ClusterConfiguration = ConvertFrom-Json ((Get-Content "c:\k\kubeclusterconfig.json" -ErrorAction Stop) | out-string)
 
+$global:$IsSkipCleanupNetwork = [System.Convert]::ToBoolean($Global:ClusterConfiguration.Services.IsSkipCleanupNetwork)
+
+filter Timestamp { "$(Get-Date -Format o): $_" }
+
+function Write-Log($message) {
```
Comment (AbelHu) [inline staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:9]: cleanupnetwork.ps1 is executed in staging/cse/windows/provisioningscripts/windowsnodereset.ps1. could we wrapper this as a function and import it in staging/cse/windows/provisioningscripts/windowsnodereset.ps1 so do not use a new log file?

Comment (ShiqianTao) [reply staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:9]: redirect to nodereset.log

File: staging/cse/windows/provisioningscripts/cleanupnetwork.ps1 (hunk 1 above)
Comment (ShiqianTao) [reply staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:68]: The cleanup process for `HNS Network` and `CNI data` should be consistent.

File: staging/cse/windows/kubernetesfunc.ps1 (hunk 10 above)
Comment (ShiqianTao) [reply staging/cse/windows/kubernetesfunc.ps1]: It is a testing code. Updated to the official one.

File: staging/cse/windows/debug/hns.v2.psm1 (hunk 9 above)
Comment (ShiqianTao) [reply staging/cse/windows/debug/hns.v2.psm1]: Sync by https://github.com/Azure/AgentBaker/pull/4279

File: staging/cse/windows/provisioningscripts/cleanupnetwork.ps1 [hunk 12]
```diff
@@ -29,16 +39,38 @@ if ($hnsNetwork) {
...
+        Get-HnsEndpoint | Remove-HnsEndpoint
+        Remove-HnsNetwork $hnsNetwork
+
+        Write-Log  "Count actively reserved port pools to log."
+        $countPortPools = 1
+        # Guarantee (10s <= wait time <= 60s) between HNS removal vs. creation for safety.
+        # Usually, it takes 10s to sleep like before, but it may take longer in some cases (e.g., COSMIC).
+        $maxIteration=6
+        For ($i=1; ($i -le $maxIteration -and $countPortPools -ne 0); $i++) {
+            $countPortPools = (Invoke-HnsRequest -Method "GET" -Type "portpools" | Select-Object -ExpandProperty PortPoolAllocations | Where-Object -FilterScript {$_.RequiresHostPortReservation -eq "True" -and $_.PortRanges} | Select-Object -ExpandProperty PortRanges).Count
```
Comment (AbelHu) [inline staging/cse/windows/provisioningscripts/cleanupnetwork.ps1:53]: do we need to use try/catch to avoid unexpected error in this block?

Comment (ShiqianTao) [reply staging/cse/windows/provisioningscripts/cleanupnetwork.ps1]: Agree. Updated code and test:
image

**Review by AbelHu — APPROVED** [review]
LGTM. Thanks for the PR.

**Pull request merged**